### ISSUE TITLE
marginalia--full-candidate: Candidate string is invalid for partial-completion

### ISSUE THREAD

**Issue by minad** [minad/marginalia#65]
When completing file names with partial completion, e.g., "~/.c/e/i", the full candidate computed by `marginalia--full-candidate` is wrong.

**Comment by oantolin** [comment]
This bug is a little tricky. As you say, we really want the base position returned by `completion-all-completions`, but without having to compute all completions. I checked the Emacs source for two places I'm aware that can correctly handle `partial-completion`'s paths: (1) pressing tab in the minibuffer under default completion, and (2) pressing RET in the completions buffer.

Both ultimately compute all completions! The relevant function in (2) is `choose-completion` which uses the variable `completion-base-position` which is set by `minibuffer-completion-help` using `completion-all-completions` ---this occurs when the completions buffer is created.

In (1) the function called is `minibuffer-complete`. You have to follow the chain of calls a bit, but this ultimately uses `completion-try-completion` (rather than `completion-all-completions`) to attempt to complete the minibuffer input. Now, the try-completion function for `partial-completion` is `completion-pcm-try-completion`, which also computes all completions and find a common prefix.

So, basically the two things in Emacs that I was aware of that handle these `partial-completion` paths correctly, both compute all completions. I guess if we converted the file annotator to an affixation function, we could get away with computing the completion base position only once. As it is, I think your `/`-counting hack is probably fine.

**Comment by minad** [comment]
Yes, this is a "bug" in the completing read api. There should be some possibility to query the base without computing all candidates. I think the point is that in order to compute the base, the completion style may have to compute all candidates in any case, which lead the original designers to the decision to throw the candidate and base computation together. And in order to save precious memory they decided to put the base into the last, which is brillant :sunglasses: 

I think the counting hack is probably the right way to go - I didn't commit it yet since I am not sure it is really correct - there is 1, + and -, maybe I made an off-by-one error?

**Comment by oantolin** [comment]
> And in order to save precious memory they decided to put the base into the last, which is brillant 😎

:D I think it used to be stored in a separate variable, actually. Maybe that was even `completion-base-position`, which now has a reduced role.

> I think the counting hack is probably the right way to go - I didn't commit it yet since I am not sure it is really correct - there is 1, + and -, maybe I made an off-by-one error?

I'll check.

**Comment by minad** [comment]
> :D I think it used to be stored in a separate variable, actually. Maybe that was even completion-base-position, which now has a reduced role.

Yes, I have seen this variable and when I saw it I had a glimmer of hope...

**Comment by minad** [comment]
> I'll check.

The question is also if there can be some other scenarios where the assumptions I made break. I should generally test this a bit better.

**Comment by oantolin** [comment]
There were indeed a couple of edge cases that your fix didn't handle, so I pushed a commit to the fix-65 branch to address them. 95b48db4f2f70addd743e3b44a771a294d0ca141

I think that should now be solid enough for the main branch, but I'll let you do some more testing if you want first.

**Comment by minad** [comment]
@oantolin Great! Thank you for looking into this. I will give it another test later on and merge it.
